### ISSUE TITLE
Add support for uploading and recording videos

### ISSUE THREAD

**Issue by nicolas-raoul** [commons-app/apps-android-commons#4]
including transcoding to Wikimedia-accepted format

Warning: We have to pay special attention to quality (on-topic, encyclopedic, good technical quality). A [2021 photo+video contest](https://commons.wikimedia.org/wiki/Commons:Share_Your_Struggle_2021/Winners) ended up with hundreds of high-quality photos but only one high-quality video.

**Comment by nicolas-raoul** [comment]
Commons only accepts WebM (VP8, VP9) and Ogg Theora videos.
This converter could be useful if converting on Android proves inconvenient: https://tools.wmflabs.org/videoconvert/index.php

**Comment by misaochan** [comment]
@brion has offered to help implement this, yay! :)

We should talk about how to best handle potential video-selfie-pocalyptic issues, too. I personally think that it would suffice to have a simple addition to the tutorial or a tooltip, when the feature is implemented... but am open to other suggestions.

**Comment by brion** [comment]
My rough plan is to test on-device transcoding to VP8 or VP9 since that can make use of the device OS and hardware codecs -- Android provides most of the plumbing needed. Performance will make or break this, but I think it should be adequate...

It might be possible to combine transcode and upload into a single pass, but I'm not sure of that (depends on if we need the final file size before we start, I forget).

I'm busy with a conference the next couple weeks; will start experimenting with the Android transcoding mid February.

**Comment by WikiDocJames** [comment]
Having the ability to upload video directly from my phone using the commons app would be wonderful. Just took some video of leaping benny, a fish, and the still images do not do them justice.

**Comment by ashishkumar468** [comment]
My bad, Thanks @maskaravivek, I had accidentally closed this one

**Comment by jidanni** [comment]
All I know is I have a
```
intent://media/external/video/media/212369#Intent;scheme=content;type=video/*;launchFlags=0x13080000;S.android.intent.extra.TITLE=V_20201117_093305_vHDR_On%201.mp4;end 
------------ 
ACTION: android.intent.action.VIEW 
DATA: content://media/external/video/media/212369 
MIME: video/* 
URI: intent://media/external/video/media/212369#Intent;scheme=content;type=video/*;launchFlags=0x13080000;S.android.intent.extra.TITLE=V_20201117_093305_vHDR_On%201.mp4;end 
FLAGS: 
FLAG_RECEIVER_FOREGROUND 
FLAG_ACTIVITY_FORWARD_RESULT 
FLAG_ACTIVITY_PREVIOUS_IS_TOP 
FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET 
EXTRAS: 
1 Class: java.lang.String 
Key: android.intent.extra.TITLE 
Value: V_20201117_093305_vHDR_On 1.mp4 

------------ 
MATCHING ACTIVITIES: 
Video player (com.asus.gallery - com.asus.gallery.app.MovieActivity) 
VLC (org.videolan.vlc - org.videolan.vlc.StartActivity) 
```
that my phone makes, that I want to upload directly, without even knowing what formats are even all about.

**Comment by jidanni** [comment]
For instance, with YouTube no matter what format we upload in, it is not exposed to the user. 

And when people view the video on YouTube, it is usually not in the same format that it was uploaded in anyway. Therefore we see websites take care of all this legal stuff for the user automatically.

**Comment by misaochan** [comment]
Hi @brion , any chance you are still interested in working on this? We've received quite a few requests for video support recently. :)

**Comment by Rishavgupta12345** [comment]
@nicolas-raoul @4D17Y4  According to https://github.com/commons-app/commons-app-documentation/blob/master/android/Students.md can i consider this project to work on in GSOC 2022 ?

**Comment by nicolas-raoul** [comment]
@Rishavgupta12345 If for any reason you are strongly motivated by this issue, you can choose it. But otherwise I would suggest #4764. You can also submit a proposal for each.

**Comment by OpenGreenStreet** [comment]
It would be really very nice if there was a solution for this!

Is there a timeline for this development?

**Comment by misaochan** [comment]
Unfortunately not. :( If anyone is interested in taking this up, please let us know and we would be more than happy to provide support!

**Comment by prototyperspective** [comment]
I think first of all it would need to be able to show videos in the search results and to play them. Really sad to see this is still not possible with the app, it's a core feature of WMC to be able to watch videos there. Issue: #5622  I don't think it makes sense to enable uploading before watching videos is made possible.

**Comment by OpenGreenStreet** [comment]
Yes, there is something to that: if I upload a video, I want to be able to see the result.

**Comment by chrisdebian** [comment]
Wow, this issue is 9 years old ;-) . @bvibber Hi, Brooke, is this something you can help with, it looked like you had some useful input, above?

Thanks,

Chris.

**Comment by OpenGreenStreet** [comment]
It would be really great if this very old feature request were implemented.
If even large MP4 files (including videos of up to 29 minutes recorded with a smartphone) were reliably converted when uploading. But that is likely to be a huge challenge.

Even with this, there are always crashes:

https://tools.wmflabs.org/video2commons/

https://tools.wmflabs.org/videoconvert

In the end, I've always had to convert locally (pc) and then upload via the Upload Wizzard.

**Comment by WikiDocJames** [comment]
Yah it would be nice to have MP4 uploads built into the upload wizard with the file format conversion just taken care of and no user every needing to learn about this issue...

**Comment by prototyperspective** [comment]
I wonder why videoconvert crashes. One could also [convert videos locally within the app](https://github.com/arthenica/ffmpeg-kit) but I don't know how long it would take with a normal smartphone.  
Making mp4 uploads possible via auto-conversion also on PC would be the optimal solution (note that low-resolution video uploads directly from smartphones isn't particularly the most useful needed kind of files), I think WMC has a lot of potential as a video-platform if it was 1. better indexed in Web search engines ([which currently semi-censor videos on Commons, not showing them in their Videos tab](https://meta.wikimedia.org/wiki/Community_Wishlist/Wishes/Do_something_about_Google_&_DuckDuckGo_search_not_indexing_media_files_and_categories_on_Commons)) and 2. if playing videos within the app (and maybe also [NewPipe](https://github.com/TeamNewPipe/NewPipe)) would work as well – new issue about that here: #5983 (3. also needed is displaying both file-title and caption not just the latter which often is only complementary to the filetitle or not meant to be a title; #6023).

**Comment by OpenGreenStreet** [comment]
@prototyperspective 

This is the message I keep getting when converting large videos: https://commons.wikimedia.org/wiki/File:Video2commons.png
It just doesn't happen anymore ...

**Comment by prototyperspective** [comment]
@OpenGreenStreet This seems unrelated to this issue; video2commons is for uploading videos already on the Web not from your device. I think the issue in the screenshot has been solved already, it just often takes a while now but if it's not uploaded by next year I suggest you make a post on the video2commons talk page.
